updater/lib/dependabot/dependency_snapshot.rb

```diff
@@ -69,9 +69,16 @@ def groups
       @dependency_group_engine.dependency_groups
     end
 
+    def calculate_ungrouped_dependencies(all_grouped_changes)
+      @ungrouped_dependencies = allowed_dependencies.select do |dep|
+        all_grouped_changes.none? { |change| change.name == dep.name }
+      end
+    end
+
     def ungrouped_dependencies
       # If no groups are defined, all dependencies are ungrouped by default.
       return allowed_dependencies unless groups.any?
+      return @ungrouped_dependencies if defined?(@ungrouped_dependencies)
 
       @dependency_group_engine.ungrouped_dependencies
```

Comment:
Can we not always store the `upgrouped_dependencies` on the `dependency_group_engine` 🤔

Reply:
I can try pushing it down into the engine, but there's no reference to dependency_group_engine in GroupUpdateAllVersions, so it will pass through the dependency_snapshot anyway or I could add it to attr_reader. 

Not sure if that's better 🤷

Reply:
I was thinking that way we could get rid of the `return @ungrouped_dependencies if defined?(@ungrouped_dependencies)`. I'm also not sure if it's better, I'm also fine keeping things as is

Reply:
Let's leave it for now, I'm considering some refactorings that we can try after this goes in that might improve the situation. Probably good to leave that for a new PR.